### PR TITLE
fix ema tests

### PR DESCRIPTION
**Pull request by sandylaker**
Signed-off-by: sandylaker <yawei.li@tum.de>

Fixes EMA GPU tests. See [#2098](https://github.com/pytorch/ignite/pull/2098#issuecomment-877491173)

Description:

Check list:

- [x] New tests are added (if a new feature is added)
- [x] New doc strings: description and/or example code are in RST format
- [x] Documentation is updated (if required)

### CODE REVIEW

File: tests/ignite/handlers/test_ema_handler.py [hunk 1]
```diff
@@ -28,7 +35,10 @@ def _get_dummy_step_fn(model: Union[nn.Module, DataParallel, DistributedDataPara
 
     def step_fn(engine, batch):
         """Increment the weight by 1 at each iteration"""
-        model.weight.data.add_(1)
+        if isinstance(model, (DataParallel, DistributedDataParallel)):
+            model.module.weight.data.add_(1)
+        else:
+            model.weight.data.add_(1)
```
Comment (vfdev-5) [inline tests/ignite/handlers/test_ema_handler.py:41]: Maybe, we can do
```suggestion
        _unwrap_model(model).weight.data.add_(1)
```
?

File: tests/ignite/handlers/test_ema_handler.py [hunk 2]
```diff
@@ -238,14 +252,14 @@ def _test_ema_final_weight(model, device, ddp=False, interval=1):
 
 
 @pytest.mark.parametrize("interval", [1, 2])
-def test_ema_final_weight_cpu(get_dummy_model, interval):
+def test_cpu(get_dummy_model, interval):
```
Comment (vfdev-5) [inline tests/ignite/handlers/test_ema_handler.py:255]:
```suggestion
def test_ema_final_weight_cpu(get_dummy_model, interval):
```
I think the name was OK.

File: tests/ignite/handlers/test_ema_handler.py [hunk 3]
```diff
@@ -238,14 +252,14 @@ def _test_ema_final_weight(model, device, ddp=False, interval=1):
...
     device = torch.device("cpu")
     _test_ema_final_weight(get_dummy_model(), device=device, ddp=False, interval=interval)
 
 
 @pytest.mark.parametrize("interval", [1, 2])
 @pytest.mark.skipif(not torch.cuda.is_available(), reason="Skip if no GPU")
-def test_ema_final_weight_cuda(get_dummy_model, interval):
+def test_cuda(get_dummy_model, interval):
```
Comment (vfdev-5) [inline tests/ignite/handlers/test_ema_handler.py:262]:
```suggestion
def test_ema_final_weight_cuda(get_dummy_model, interval):
```
same here.

**Review by vfdev-5 — COMMENTED** [review]
Thanks for the updates @sandylaker 
I proposed few other minor suggestions
Meanwhile, I'll open another PR from this PR to execute GPU test.

File: tests/ignite/handlers/test_ema_handler.py [hunk 4]
```diff
@@ -273,7 +283,7 @@ def test_ema_final_weight_gloo_cuda(get_dummy_model, distributed_context_single_
 @pytest.mark.skipif(not idist.has_hvd_support, reason="Skip if no Horovod dist support")
 @pytest.mark.skipif("WORLD_SIZE" in os.environ, reason="Skip if launched as multiproc")
 @pytest.mark.skipif(not torch.cuda.is_available(), reason="Skip if no GPU")
-def test_ema_final_weight_hvd_cuda(gloo_hvd_executor, interval):
+def test_ema_final_weight_distrib_hvd_cuda(gloo_hvd_executor, interval):
```
Comment (vfdev-5) [inline tests/ignite/handlers/test_ema_handler.py:286]:
```suggestion
def test_ema_final_weight_distrib_hvd(gloo_hvd_executor, interval):
```

File: tests/ignite/handlers/test_ema_handler.py [hunk 5]
```diff
@@ -314,6 +324,6 @@ def test_ema_final_weight_multinode_gloo_cuda(get_dummy_model, distributed_conte
 @pytest.mark.parametrize("interval", [1, 2])
 @pytest.mark.skipif(not idist.has_native_dist_support, reason="Skip if no native dist support")
 @pytest.mark.skipif("GPU_MULTINODE_DISTRIB" not in os.environ, reason="Skip if not multi-node distributed")
-def test_ema_final_weight_multinode_nccl_cuda(get_dummy_model, distributed_context_multi_node_nccl, interval):
+def test_ema_final_weight_distrib_multinode_nccl_cuda(get_dummy_model, distributed_context_multi_node_nccl, interval):
```
Comment (vfdev-5) [inline tests/ignite/handlers/test_ema_handler.py:327]:
```suggestion
def test_ema_final_weight_distrib_multinode_nccl_gpu(get_dummy_model, distributed_context_multi_node_nccl, interval):
```

File: tests/ignite/handlers/test_ema_handler.py [hunk 6]
```diff
@@ -305,7 +315,7 @@ def _test_ema_final_weight_xla_nprocs(index):
 @pytest.mark.parametrize("interval", [1, 2])
 @pytest.mark.skipif(not idist.has_native_dist_support, reason="Skip if no native dist support")
 @pytest.mark.skipif("MULTINODE_DISTRIB" not in os.environ, reason="Skip if not multi-node distributed")
-def test_ema_final_weight_multinode_gloo_cuda(get_dummy_model, distributed_context_multi_node_gloo, interval):
+def test_ema_final_weight_distrib_multinode_gloo_cuda(get_dummy_model, distributed_context_multi_node_gloo, interval):
```
Comment (vfdev-5) [inline tests/ignite/handlers/test_ema_handler.py:318]:
```suggestion
def test_ema_final_weight_distrib_multinode_gloo_cpu_or_gpu(get_dummy_model, distributed_context_multi_node_gloo, interval):
```

File: tests/ignite/handlers/test_ema_handler.py [hunk 7]
```diff
@@ -314,6 +324,6 @@ def test_ema_final_weight_multinode_gloo_cuda(get_dummy_model, distributed_conte
 @pytest.mark.parametrize("interval", [1, 2])
```
Comment (vfdev-5) [inline tests/ignite/handlers/test_ema_handler.py:324]:
```suggestion
@pytest.mark.multinode_distributed
@pytest.mark.parametrize("interval", [1, 2])
```

File: tests/ignite/handlers/test_ema_handler.py [hunk 8]
```diff
@@ -273,25 +283,25 @@ def test_ema_final_weight_gloo_cuda(get_dummy_model, distributed_context_single_
 @pytest.mark.skipif(not idist.has_hvd_support, reason="Skip if no Horovod dist support")
 @pytest.mark.skipif("WORLD_SIZE" in os.environ, reason="Skip if launched as multiproc")
 @pytest.mark.skipif(not torch.cuda.is_available(), reason="Skip if no GPU")
-def test_ema_final_weight_hvd_cuda(gloo_hvd_executor, interval):
+def test_ema_final_weight_distrib_hvd(gloo_hvd_executor, interval):
     device = torch.device("cuda")
     nproc = torch.cuda.device_count()
```
Comment (vfdev-5) [inline tests/ignite/handlers/test_ema_handler.py:288]:
```suggestion
    device = torch.device("cpu" if not torch.cuda.is_available() else "cuda")
    nproc = 4 if not torch.cuda.is_available() else torch.cuda.device_count()```

File: tests/ignite/handlers/test_ema_handler.py [hunk 9]
```diff
@@ -273,25 +283,25 @@ def test_ema_final_weight_gloo_cuda(get_dummy_model, distributed_context_single_
 @pytest.mark.skipif(not idist.has_hvd_support, reason="Skip if no Horovod dist support")
 @pytest.mark.skipif("WORLD_SIZE" in os.environ, reason="Skip if launched as multiproc")
 @pytest.mark.skipif(not torch.cuda.is_available(), reason="Skip if no GPU")
-def test_ema_final_weight_hvd_cuda(gloo_hvd_executor, interval):
-    device = torch.device("cuda")
-    nproc = torch.cuda.device_count()
+def test_ema_final_weight_distrib_hvd(gloo_hvd_executor, interval):
+    device = torch.device("cpu" if not torch.cuda.is_available() else "cuda")
+    nproc = 4 if not torch.cuda.is_available() else torch.cuda.device_count()```
```
Comment (vfdev-5) [inline tests/ignite/handlers/test_ema_handler.py:288]:
```suggestion
    device = torch.device("cpu" if not torch.cuda.is_available() else "cuda")
    nproc = 4 if not torch.cuda.is_available() else torch.cuda.device_count()
```

**Comment by sandylaker** [comment]
@vfdev-5 It seems that [badba40](https://github.com/pytorch/ignite/pull/2102/commits/badba402be26f57a90daa4e048401cb1b72cc3d7) passed all the GPU tests except for the two horovod distributed tests. I cannot see the error message since it is cancelled. Could you please help me to check it out? Thanks a lot.

**Comment by vfdev-5** [comment]
@sandylaker 
here is the error message:
```
_____________________________________________________________________ test_ema_final_weight_distrib_hvd[1] ______________________________________________________________________
Fixture "get_dummy_model" called directly. Fixtures are not meant to be called directly,                                                                                         
but are created automatically when test functions request them as parameters.                                                                                                    
See https://docs.pytest.org/en/stable/fixture.html for more information about fixtures, and  
```

**Comment by vfdev-5** [comment]
To reproduce the issue you can use this docker image:
```
docker run -it --name=ignite-dev-hvd -v $PWD:/ignite -w /ignite --network=host --shm-size 16G pytorchignite/hvd-base:1.9.0-0.4.5 /bin/bash

$ pip install -r requirements-dev.txt
$ CUDA_VISIBLE_DEVICES=0 pytest -vvv tests/ -m distributed -k test_ema_final_weight_distrib_hvd
```

File: tests/ignite/handlers/test_ema_handler.py [hunk 10]
```diff
@@ -273,25 +283,25 @@ def test_ema_final_weight_gloo_cuda(get_dummy_model, distributed_context_single_
 @pytest.mark.skipif(not idist.has_hvd_support, reason="Skip if no Horovod dist support")
 @pytest.mark.skipif("WORLD_SIZE" in os.environ, reason="Skip if launched as multiproc")
 @pytest.mark.skipif(not torch.cuda.is_available(), reason="Skip if no GPU")
```
Comment (vfdev-5) [inline tests/ignite/handlers/test_ema_handler.py:285]:
```suggestion
```

File: tests/ignite/handlers/test_ema_handler.py [hunk 11]
```diff
@@ -224,8 +234,9 @@ def _test_ema_final_weight(model, device, ddp=False, interval=1):
     # engine will run 4 iterations
     engine.run(range(2), max_epochs=2)
 
-    ema_weight = ema_handler.ema_model.weight.data
-    model_weight = model.weight.data
+    # ema_model and model can be DP or DDP
+    ema_weight = _unwrap_model(ema_handler.ema_model).weight.data
+    model_weight = _unwrap_model(model).weight.data
     assert ema_weight.device == device
     assert model_weight.device == device
```
Comment (vfdev-5) [inline tests/ignite/handlers/test_ema_handler.py:241]:
```suggestion
    assert ema_weight.device.type == device.type
    assert model_weight.device == ema_weight.device
```

**Comment by vfdev-5** [comment]
@sandylaker thanks for the fixes, it should be passing now.

File: tests/ignite/handlers/test_ema_handler.py (hunk 11 above)
Comment (sandylaker) [reply tests/ignite/handlers/test_ema_handler.py:244]: The problem is that in the horovod test, the processes will be spawn by the horovod executor. Thus, if we pass `device = "cuda"` to the function, `model = idist.auto_model(model)` will move the `model` to `idist.device()`, which might be `f"cuda:{idist.local_rank()}"`, however `device` is `torch.device("cuda")`. In order to fix this, I change the signature of `_test_ema_final_weight` to `_test_ema_final_weight(model, device=None, ddp=False, interval=1)`. If `device=None`, we call `idsti.device()` to get the current device. Then, we explicitly pass `None` to this function in the horovod executor. I have run the tests on my server with single node, and 22 passed, 6 skipped.

Comment (vfdev-5) [reply tests/ignite/handlers/test_ema_handler.py:244]: Yes, your solution will work as well and looks good to me. Thanks !

**Review by vfdev-5 — APPROVED** [review]
LGTM! Thanks @sandylaker

**Pull request merged**